### ISSUE TITLE
Decimal places for PriceAmount fixed at 2 digits - need 4 digits e.g. for fuel invoice

### ISSUE THREAD

**Issue by harzmann** [stephanstapel/ZUGFeRD-csharp#398]
PriceAmount is formatted or rounded to 2 decimal places. However, for example, for fuel bills, 4 decimal places are required for the price per liter. Temporary Save function modified: 
Writer.WriteValue(_formatDecimal(tradeLineItem.GrossUnitPrice.Value,4));
![image](https://github.com/user-attachments/assets/fd33b3e5-ccdd-4c5f-9d59-756a1803b27b)

**Comment by stephanstapel** [comment]
Thanks for your suggestion. I'd be more than happy if you could create a PR. I propose to format the decimal with two digits in case the actual value has two or less digits and use four digits only in case they are really set in the decimal value.

Please add a test case to https://github.com/stephanstapel/ZUGFeRD-csharp/blob/master/ZUGFeRD-Test/XRechnungUBLTests.cs

**Comment by HolgerPfeifer** [comment]
And the same would be useful for the AllowanceCharge.Amount.

**Comment by stephanstapel** [comment]
@HolgerPfeifer : would you like to contribute?

**Comment by HolgerPfeifer** [comment]
Yes, I would like. It will take me some time to get my devolopment environment ready for GitHub. Currently I only use local BitBucket repositories.

**Comment by stephanstapel** [comment]
Hi @ernie1000 ,

This change was made intentionally and cannot be 'just reversed' as this component is part of lots of applications.
I am happy if you send in a pull request to add digits on demand as I suggested above,

**Comment by wh1980** [comment]
In our invoicing software, the user can even use values with up to 10 decimals.
It would be essential to not round the values as it distorts the invoice and inconsistencies and contradictions occur.
Just leave it up to the user if he passes values with 2, 4 or even 10 decimals.
It would be good to only include them into the xml code as they are and don't change them.

The idea of @ernie1000 is good:
` TradeLineItem.AmountDecimals=4; `

**Comment by DaveHalma** [comment]
Be careful when changing the numDecimals (IInvoiceDescriptorWriter) to 4 or higher, thus leading to many errors when validating. Though, you can change certain amounts, like the “Unit Price Amount” (see semantics). Just add the desired value as a parameter like this to your Writer/s:
                    Writer.WriteStartElement("cbc”, “BaseAmount”); // BT-148
                    Writer.WriteAttributeString(“currencyID”, this.Descriptor.Currency.EnumToString());
                    Writer.WriteValue(_formatDecimal(charges[0].BasisAmount, 4)); // ADD VALUE HERE
Please feel free to correct me!

**Comment by DaveHalma** [comment]
Be careful when changing “formatDecimal” in “InvoiceDescriptorWriter” thus this can lead to many errors when validating.

**Comment by stephanstapel** [comment]
We should rely on the standard:
https://github.com/stephanstapel/ZUGFeRD-csharp/blob/master/documentation/zugferd23en/Documentation/1.%20FACTUR-X%201.0.07%20EN.pdf

i.e.

![grafik](https://github.com/user-attachments/assets/fc643c86-a34b-4dc5-bd37-35cd492f4181)

It would be great if you could check the writers if it matches these decimal rules for all versions and formats.

**Comment by DaveHalma** [comment]
The writers work great for me, but I needed to adjust “Unit Price Amount” and “Quantity” manually to 4 due to some rounding issues. I think it should be left to anyone to adjust the size of the numbers behind the decimal separator, as this seems quite individual. On the other hand, we could implement the maximum of allowed decimals just for accessibility. Your call.

**Comment by wh1980** [comment]
The validators are not complaining when using values with 4 decimals in GrossPriceProductTradePrice and also in NetPriceProductTradePrice. We need it like that as our users can switch between 2 and 4 decimals for the unit price amounts:

`      <ram:SpecifiedLineTradeAgreement>
        <ram:GrossPriceProductTradePrice>
          <ram:ChargeAmount>9.9925</ram:ChargeAmount>
          <ram:BasisQuantity unitCode="MMT">1.0000</ram:BasisQuantity>
          <ram:AppliedTradeAllowanceCharge>
            <ram:ChargeIndicator>
              <udt:Indicator>false</udt:Indicator>
            </ram:ChargeIndicator>
            <ram:ActualAmount>2.7502</ram:ActualAmount>
          </ram:AppliedTradeAllowanceCharge>
        </ram:GrossPriceProductTradePrice>
        <ram:NetPriceProductTradePrice>
          <ram:ChargeAmount>7.2423</ram:ChargeAmount>
          <ram:BasisQuantity unitCode="MMT">1.0000</ram:BasisQuantity>
        </ram:NetPriceProductTradePrice>
      </ram:SpecifiedLineTradeAgreement>`

**Comment by wh1980** [comment]
The validators are not complaining when using values with 4 decimals in **GrossPriceProductTradePrice\ChargeAmount**, **GrossPriceProductTradePrice\AppliedTradeAllowanceCharge\ActualAmount** and also in **NetPriceProductTradePrice\ChargeAmount**. We need it like that as our users can switch between **2 and 4 decimals** for the unit price amounts:

`<ram:SpecifiedLineTradeAgreement>
        <ram:GrossPriceProductTradePrice>
          <ram:ChargeAmount>9.9925</ram:ChargeAmount>
          <ram:BasisQuantity unitCode="MMT">1.0000</ram:BasisQuantity>
          <ram:AppliedTradeAllowanceCharge>
            <ram:ChargeIndicator>
              <udt:Indicator>false</udt:Indicator>
            </ram:ChargeIndicator>
            <ram:ActualAmount>2.7502</ram:ActualAmount>
          </ram:AppliedTradeAllowanceCharge>
        </ram:GrossPriceProductTradePrice>
        <ram:NetPriceProductTradePrice>
          <ram:ChargeAmount>7.2423</ram:ChargeAmount>
          <ram:BasisQuantity unitCode="MMT">1.0000</ram:BasisQuantity>
        </ram:NetPriceProductTradePrice>
      </ram:SpecifiedLineTradeAgreement>`

We need something like this:
 **TradeLineItem.UnitPriceAmountDecimals=4;**
